substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -1089,24 +1129,62 @@ impl<T: Config> Module<T> {
 
-        // Send 50% to the sales channel
-        let sales_share = Perbill::from_percent(50) * amount;
-        debug::info!(
-            "Transfering: {:?} from contract twin {:?} to sales account {:?}",
-            &sales_share,
-            &twin.account_id,
-            &pricing_policy.certified_sales_account
-        );
-        <T as Config>::Currency::transfer(
-            &twin.account_id,
-            &pricing_policy.certified_sales_account,
-            sales_share,
-            KeepAlive,
-        )
-        .map_err(|_| DispatchError::Other("Can't make sales share transfer"))?;
+        let mut sales_share = 50;
+
+        if let Some(provider_id) = contract.solution_provider_id {
+            let solution_provider = SolutionProviders::<T>::get(provider_id);
+            let total_take: u8 = solution_provider
+                .providers
+                .iter()
+                .map(|provider| provider.take)
+                .sum();
+            sales_share -= total_take;
+
+            if !solution_provider.providers.iter().map(|provider| {
+                let share = Perbill::from_percent(provider.take as u32) * amount;
+                debug::info!(
+                    "Transfering: {:?} from contract twin {:?} to provider account {:?}",
+                    &share,
+                    &twin.account_id,
+                    &provider.who
+                );
+                <T as Config>::Currency::transfer(
+                    &twin.account_id,
+                    &provider.who,
+                    share,
+                    KeepAlive,
+                )
+            })
+            .filter(|result| result.is_err())
+            .collect::<Vec<DispatchResult>>()
+            .is_empty() {
+                return Err(DispatchError::from(Error::<T>::InvalidProviderConfiguration))
+            }
+        };
+
+        println!("sales share: percentage {:?}", sales_share);
+
+        if sales_share > 0 {
+            let share = Perbill::from_percent(sales_share.into()) * amount;
+            // Send 50% to the sales channel
+            debug::info!(
+                "Transfering: {:?} from contract twin {:?} to sales account {:?}",
+                &sales_share,
+                &twin.account_id,
+                &pricing_policy.certified_sales_account
+            );
+            <T as Config>::Currency::transfer(
+                &twin.account_id,
+                &pricing_policy.certified_sales_account,
+                share,
+                KeepAlive,
+            )
+            .map_err(|_| DispatchError::Other("Can't make staking pool share transfer"))?;
+        }
 
         // Burn 35%, to not have any imbalance in the system, subtract all previously send amounts with the initial
-        let mut amount_to_burn = amount - foundation_share - staking_pool_share - sales_share;
+        let mut amount_to_burn =
+            amount - foundation_share - staking_pool_share - (Perbill::from_percent(50) * amount);
```

Comment:
multiply the first `amount`, then subtract the foundation and staking pool shares. That way the treasury amount is implicitly deducted, which saves a subtraction.